### ISSUE TITLE
Manipulating path segments

### ISSUE THREAD

**Issue by SimonSapin** [servo/rust-url#188]
`Url` currently has `pop_path_segments` and `push_path_segment` methods, but they have a design problem reported by @sfackler. A path is at minimum `/` which is considered like a single path segment that happens to be empty (like trailing slashes in general), so `push_path_segment("foo")` make the path be `//foo` with two slashes.

Fixing this may involve breaking API changes, so instead I’ll make these methods private (effectively removing them) before publishing 1.0.0, in order to unblock it. I’ll add them back (or something else to provide this functionality) in 1.1.0.

**Comment by sfackler** [comment]
In the meantime, what's the right way to add path segments? Something like
```
let mut url = ...;
url.set_path(&format!("{}/{}", url.path(), percent_encode(new_component.as_bytes(), PATH_SEGMENT_ENCODE_SET);
```
?

**Comment by sfackler** [comment]
I guess it would actually need to check if `url.path() == "/"` to avoid the empty path problem.

**Comment by SimonSapin** [comment]
Yes, something like that.

**Comment by maximih** [comment]
I guess the current issue is related so I am posting here. Let me know if I should open a new issue.

I am trying to switch to `url 1.0.0` but I cannot find the equivalent of the `url 0.5` code below:

```rust
let mut path = url.path_mut().unwrap();
path.push(fragment.to_owned());
```

I am trying to append a segment to a path which does end with a trailing `/` and `url.join()` cannot handle this scenario. Is there a way to do this with `url 1.0.0` or I need to do some ugly string concat to make sure a trailing ´/´ is always present and then use `url.join()`?

Thanks

**Comment by SimonSapin** [comment]
In 1.0.0 you have to use string concatenation and `url.set_path(p)`. In 1.1.0 I want to add something like `url.push_path_segment(s)` (but maybe a different API) which would do the right thing in various situations.

**Comment by jgillich** [comment]
The example by @sfackler doens't appear to work, you'd have to create a `String` out of `url.path()` first, which is not great.
```
src/transport.rs:46:40: 46:43 error: cannot borrow `url` as immutable because it is also borrowed as mutable [E0502]
src/transport.rs:46         url.set_path(&format!("{}/{}", url.path(), path));
                                                           ^~~
<std macros>:2:26: 2:57 note: in this expansion of format_args!
src/transport.rs:46:23: 46:57 note: in this expansion of format! (defined in <std macros>)
src/transport.rs:46:9: 46:12 note: previous borrow of `url` occurs here; the mutable borrow prevents subsequent moves, borrows, or modification of `url` until the borrow ends
src/transport.rs:46         url.set_path(&format!("{}/{}", url.path(), path));
                            ^~~
src/transport.rs:46:58: 46:58 note: previous borrow ends here
src/transport.rs:46         url.set_path(&format!("{}/{}", url.path(), path));
```

**Comment by SimonSapin** [comment]
Yeah, for now you have to move the `format!(…)` expression into a separate variable.

The API I have in mind would be something like this:

```rust
impl Url {
    fn path_segments_mut(&mut self) -> PathSegmentsMut { … }
}
struct PathSegmentsMut<'a> {
